sdk/nodejs/stackReference.ts

```diff
@@ -112,16 +124,67 @@ export class StackReference extends CustomResource {
      * @param name The name of the stack output to fetch.
      */
     public requireOutputSync(name: string): any {
-        const out = this.requireOutput(name);
-        const isSecret = promiseResult(out.isSecret);
-        if (isSecret) {
-            throw new Error("Cannot call 'requireOutputSync' if the referenced stack has secret outputs. Use 'requireOutput' instead.");
+        const [outputs, secretNames] = this.readOutputsSync("requireOutputSync");
+
+        if (secretNames.indexOf(name) >= 0) {
+            throw new Error("Cannot call 'requireOutputSync' if the referenced stack output is a secret. Use 'requireOutput' instead.");
+        }
+
+        if (!outputs.hasOwnProperty(name)) {
+            throw new Error(`Required output '${name}' does not exist on stack.`);
+        }
+
+        return outputs[name];
+    }
+
+    private readOutputsSync(callerName: string): [Record<string, any>, string[]] {
+        // See if we already read in the outputs synchronously.  If so, just use those values.
+        if (this.syncOutputs) {
+            return [this.syncOutputs, this.syncSecretOutputNames!];
+        }
+
+        // We need to pass along our StackReference name to the engine so it knows what results to
+        // return.  However, because we're doing this synchronously, we can only do this safely if
+        // the stack-reference name is synchronously known (i.e. it's a string and not a
+        // Promise/Output).  If it is only asynchronously known, then warn the user and make an unsafe
+        // call to the deasync lib to get the name.
+        let stackName: string;
+        if (this.stackReferenceName instanceof Promise) {
+            // Have to do an explicit console.log here as the call to utils.promiseResult may hang
+            // node, and that may prevent our normal logging calls from making it back to the user.
+            console.log(
+                `Call made to StackReference.${callerName} with a StackReference with a Promise name. This may cause the program to hang.
+                For more details see: https://www.pulumi.com/docs/troubleshooting/#stackreference-sync`);
+
+            stackName = promiseResult(this.stackReferenceName);
+        }
+        else if (Output.isInstance(this.stackReferenceName)) {
+            console.log(
+                `Call made to StackReference.${callerName} with a StackReference with an Output name. This may cause the program to hang.
+                For more details see: https://www.pulumi.com/docs/troubleshooting/#stackreference-sync`);
+
+            stackName = promiseResult(this.stackReferenceName.promise());
+        }
+        else {
+            stackName = this.stackReferenceName;
         }
 
-        return promiseResult(out.promise());
+        const res = invoke.invokeSync<ReadStackOutputsResult>(
+            "pulumi:pulumi:readStackOutputs", { name: stackName });
```

Comment:
Is it possible for this to fail because the CLI is older than the SDK, leading to this API not being available?  Do we need to fall back to something else?  Or present a useful version compatibility error?

Reply:
Good point.  will liekly have to address this.